enhancements/subscription-content/subscription-injection.md

```diff
@@ -0,0 +1,397 @@
+cluster-scoped:
+
+```yaml
+kind: ClusterBundle
+apiVersion: subscription.openshift.io/v1
+metadata:
+  name: rhel7
+spec:
+  entitlement:
+    name: etc-pki-entitlement
+  yumRepository:
+    name: redhat-repo-rhel7
+```
+
+and
+
+```yaml
+kind: ClusterBundle
+apiVersion: subscription.openshift.io/v1
+metadata:
+  name: rhel8
+spec:
+  entitlement:
+    name: etc-pki-entitlement
+  yumRepository:
+    name: redhat-repo-rhel8
+```
+
+The `entitlement` object is a reference to a `Secret` in the `openshift-config` namespace containing
+the entitlement keys for the cluster's subscription. The `yumRepository` object is a reference to a
+`ConfigMap` in the `openshift-config` namespace containing the yum repository definitions for the
+desired major version of RHEL.
+
+The cluster administrator is responsible for obtaining the `entitlement` and `yumRepository`
+information and adding them to the cluster. Administrators can refer to existing guidance on how to
+obtain this information.
+
+#### Subscription Bundle Controller
+
+When a `ClusterBundle` object is created, the Subscription Bundle Controller (a new controller) will
+perform the necessary actions to make the bundle available to cluster workloads. This consists of
+the following:
+
+1. Verify that the referenced `entitlement` and `yumRepository` resources exist in the
+   `openshift-config` namespace.
+2. Create a share for the `entitlement` and `yumRepository` resources as declared in the Projected
+   Resource CSI Driver proposal.
+3. Create a `ClusterRole` associated with the bundle, which grants `get`, `watch`, and `list`
+   permissions to the associated shares.
+4. Aggregate the cluster role to a configurable list of system roles. By default this will be
+   aggregated to the `edit` and `admin` roles.
+5. Add owner references to the generated `ClusterRole` and shares. This ensures the dependent
+   objects are deleted when the `ClusterBundle` is deleted.
+
+#### Subscription Injection
+
+Developers can inject subscription bundles by adding the `subscription.openshift.io/cluster-bundle`
+annotation to a pod template spec. Any workload controller which ultimately creates a `Pod` will be
+supported. For example, to inject a subscription into a `Deployment`, use the following YAML:
+
+```yaml
+kind: Deployment
+apiVersion: v1
+metadata:
+  name: my-rhel-deployment
+spec:
+  replicas: 2
+  selector:
+    matchLabels:
+      app: "rhel-deployment"
+  template:
+    metadata:
+      annotations:
+        "subscription.openshift.io/cluster-bundle": rhel8
+      labels:
+        app: "rhel-deployment"
+    spec:
+      containers:
+      - name: run
+        image: "registry.redhat.io/ubi8/ubi:latest"
+        ...
+```
+
+The Subscription Injection mutating admission webhook will watch all newly created pods. The webhook
+will do the following:
+
+1. Look for the `subscription.openshift.io/cluster-bundle` annotation
+2. Read the `ClusterBundle` with a name matching the annotation value
```

Comment:
there be an RBAC/SAR check here to make sure the SA associated with the pod has access to the ClusterBundle, right?

Reply:
So the real question is "what is going to gate access to the shared secret?"

First line of defense is the`ClusterRole` that is associated with the ProjectedResource "share" (the custom resource for the CSI driver). This role is reported in the `status` section of the `ClusterBundle`, so admins can create their own `ClusterRoleBinding` if they have locked down an SA's permissions. The CSI driver should be responsible for the SAR check, and refuse to mount the volume if the SA doesn't have sufficient permissions to read the ProjectedResource "share".

So should we add a second line of defense with an additional SAR check in the admission webhook? We could by adding `ClusterBundle` to the list of resources in the associated cluster role. We would then run the SAR check and reject the pod outright because the SA doesn't have sufficient permissions. This is actually a very good user experience - I'd rather have, "sorry, you don't have permission to create this pod" offhand than see a pod sit `Pending` waiting for a volume that will never mount.

I've convinced myself we should add the check here.